### ISSUE TITLE
CI broken due to recently introduced pyenv bug

### ISSUE THREAD

**Issue by benjyw** [pantsbuild/pex#2021]
The last successful CI run was Dec 15 2022. Then we attempted no CI runs for about two weeks, and runs are now failing on macos-11: https://github.com/pantsbuild/pex/actions/runs/3813840945/jobs/6487802924

The reason appears to be that the pyenv-installed 2.7.18 interpreter reports its `sys.version_info` as `(2, 7, 16)`. Or at least, that is what `PythonIdentity` is seeing.

~~Notably, GitHub Actions released a new macos-11 image on Dec 16: https://github.com/actions/runner-images/releases/tag/macOS-11%2F20221215.1 (it is not clear to me what "image" means in the macos context).~~
This was a red herring.

**Comment by benjyw** [comment]
Was able to set up SSH access to the runner ( - recommended) and, shockingly:

```
Mac-1672545118518:pex runner$ .pyenv_test/pyenv/versions/2.7.18/bin/python

WARNING: Python 2.7 is not recommended. 
This version is included in macOS for compatibility with legacy software. 
Future versions of macOS will not include Python 2.7. 
Instead, it is recommended that you transition to using 'python3' from within Terminal.

Python 2.7.16 (default, Aug 29 2022, 10:34:32) 
[GCC Apple LLVM 12.0.5 (clang-1205.0.19.59.6) [+internal-os, ptrauth-isa=deploy on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

I <3 macOS...

**Comment by benjyw** [comment]
Indeed:

```
/usr/bin/python2.7
Mac-1672545118518:pex runner$ /usr/bin/python2.7

WARNING: Python 2.7 is not recommended. 
This version is included in macOS for compatibility with legacy software. 
Future versions of macOS will not include Python 2.7. 
Instead, it is recommended that you transition to using 'python3' from within Terminal.

Python 2.7.16 (default, Aug 29 2022, 10:34:32) 
[GCC Apple LLVM 12.0.5 (clang-1205.0.19.59.6) [+internal-os, ptrauth-isa=deploy on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>>
```

Presumably this was 2.7.18 on the previous runner? But either way, pyenv should not get confused like this! Maybe this is due to the unorthodox way we call it in testing.py.

**Comment by benjyw** [comment]
`actions/setup-python` is doing the right thing though:

```
Mac-1672545118518:pex runner$  /Users/runner/hostedtoolcache/Python/2.7.18/x64/bin/python
Python 2.7.18 (default, Oct 29 2020, 07:55:29) 
[GCC 4.2.1 Compatible Apple LLVM 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

**Comment by benjyw** [comment]
Can probably work around this by doing something similar to   _PEX_TEST_PYENV_VERSIONS and skipping installing 2.7 via pyenv. But I'd rather get to the bottom of this. If pyenv can point 2.7.18 to the system 2.7.16 then the world as we assume it to be is badly broken.

**Comment by benjyw** [comment]
Can reproduce this outside of any test by manually replicating our pyenv steps:

```
Mac-1672548315571:pyenv runner$ ${PYENV_ROOT}/bin/pyenv install --keep 2.7.18
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-2.7.18.tar.xz...
-> https://www.python.org/ftp/python/2.7.18/Python-2.7.18.tar.xz
Installing Python-2.7.18...
patching file configure
patching file configure.ac
patching file setup.py
patching file Mac/Tools/pythonw.c
patching file setup.py
patching file Doc/library/ctypes.rst
patching file Lib/test/test_str.py
patching file Lib/test/test_unicode.py
patching file Modules/_ctypes/_ctypes.c
patching file Modules/_ctypes/callproc.c
patching file Modules/_ctypes/ctypes.h
patching file Modules/_ctypes/callproc.c
patching file setup.py
patching file Mac/Modules/qt/setup.py
patching file setup.py
python-build: use tcl-tk from homebrew
python-build: use readline from homebrew
python-build: use zlib from xcode sdk

Installed Python-2.7.18 to /Users/runner/work/pex/pex/.pyenv_test/pyenv/versions/2.7.18
Mac-1672548315571:pyenv runner$ 
Mac-1672548315571:pyenv runner$ ls ${PYENV_ROOT}/versions
2.7.18
Mac-1672548315571:pyenv runner$ ls ${PYENV_ROOT}/versions/2.7.18
bin     include lib     share
Mac-1672548315571:pyenv runner$  ${PYENV_ROOT}/versions/2.7.18/bin/python

WARNING: Python 2.7 is not recommended. 
This version is included in macOS for compatibility with legacy software. 
Future versions of macOS will not include Python 2.7. 
Instead, it is recommended that you transition to using 'python3' from within Terminal.

Python 2.7.16 (default, Aug 29 2022, 10:34:32) 
[GCC Apple LLVM 12.0.5 (clang-1205.0.19.59.6) [+internal-os, ptrauth-isa=deploy on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

**Comment by benjyw** [comment]
After playing around with pyenv on the box, looks like the GHA image update was a red herring. If I pin pyenv to an earlier release (v2.3.8) instead of HEAD, the problem doesn't occur. Confirming in a CI run now.

**Comment by benjyw** [comment]
Confirmed. Fix in https://github.com/pantsbuild/pex/pull/2020. 

Will try and repro locally and bisect it in the pyenv repo to see if I can give them a sensible bug report (or fix).

Yeesh...

**Comment by benjyw** [comment]
Fixed in #2020

**Comment by benjyw** [comment]
Reported to pyenv here: https://github.com/pyenv/pyenv/issues/2578
